R/patch.R

```diff
@@ -0,0 +1,375 @@
+#'   c = tibble(
+#'     c1 = 1:5
+#'   ),
+#'   d = matrix(1:10, 5, 2)
+#' )
+#' tbl2 <- tbl1
+#' tbl2$a[[2]] <- tbl1$a[[2]] + 1e-10
+#' tbl2$b[[3]][[1]] <- tbl1$b[[3]][[1]] + 1e-10
+#' tbl2$c$c1[[4]] <- tbl1$c$c1[[4]] + 1e-10
+#' tbl2$d[[5, 2]] <- tbl1$d[[5, 2]] + 1e-10
+#' vctrs::vec_equal(tbl1, tbl2, na_equal = TRUE)
+#' vec_approx_equal(tbl1, tbl2, na_equal = TRUE, abs_tol = 1e-12)
+#' vec_approx_equal(tbl1, tbl2, na_equal = TRUE, abs_tol = 1e-8)
+#'
+#'
+#'
+#'
+#'
+#' # Type comparison within lists is stricter, matching vctrs:
+#' vctrs::vec_equal(list(1:2), list(as.numeric(1:2)))
+#' vec_approx_equal(list(1:2), list(as.numeric(1:2)), FALSE, abs_tol = 0)
+#'
+#' @export
+vec_approx_equal <- function(vec1, vec2, na_equal, .ptype = NULL, ..., abs_tol, inds1 = NULL, inds2 = NULL) {
+  if (!obj_is_vector(vec1)) cli_abort("`vec1` must be recognized by vctrs as a vector")
+  if (!obj_is_vector(vec2)) cli_abort("`vec2` must be recognized by vctrs as a vector")
+  # Leave vec size checking to vctrs recycling ops.
+  assert_logical(na_equal, any.missing = FALSE, len = 1L)
+  # Leave .ptype checks to cast operation.
+  check_dots_empty()
+  assert_numeric(abs_tol, lower = 0, len = 1L)
+  assert(
+    check_null(inds1),
+    check_numeric(inds1),
+    check_logical(inds1),
+    check_character(inds1)
+  )
+  assert(
+    check_null(inds2),
+    check_numeric(inds2),
+    check_logical(inds2),
+    check_character(inds2)
+  )
+  # Leave heavier index validation to the vctrs recycling & indexing ops.
+
+  # Recycle inds if provided; vecs if not:
+  common_size <- vec_size_common(
+    if (is.null(inds1)) vec1 else inds1,
+    if (is.null(inds2)) vec2 else inds2
+  )
+  if (is.null(inds1)) {
+    vec1 <- vec_recycle(vec1, common_size)
+  } else {
+    inds1 <- vec_recycle(inds1, common_size)
+  }
+  if (is.null(inds2)) {
+    vec2 <- vec_recycle(vec2, common_size)
+  } else {
+    inds2 <- vec_recycle(inds2, common_size)
+  }
+  if (!identical(vec_ptype(vec1), vec_ptype(vec2)) || !is.null(.ptype)) {
+    # perf: this is slow, so try to avoid it if it's not needed
+    vecs <- vec_cast_common(vec1, vec2, .to = .ptype)
+  } else {
+    vecs <- list(vec1, vec2)
+  }
+  vec_approx_equal0(vecs[[1]], vecs[[2]], na_equal, abs_tol, inds1, inds2)
+}
+
+#' Helper for [`vec_approx_equal`] for vecs guaranteed to have the same ptype and size
+#'
+#' @keywords internal
+vec_approx_equal0 <- function(vec1, vec2, na_equal, abs_tol, inds1 = NULL, inds2 = NULL) {
+  if (is_bare_numeric(vec1) && abs_tol != 0) {
+    # Matching vec_equal, we ignore names and other attributes.
+    if (!is.null(inds1)) vec1 <- vec_slice(vec1, inds1)
+    if (!is.null(inds2)) vec2 <- vec_slice(vec2, inds2)
+    na_or_nan1 <- is.na(vec1)
+    na_or_nan2 <- is.na(vec2)
+    # Since above are bare logical vectors, we can use `fifelse`
+    res <- fifelse(
+      !na_or_nan1 & !na_or_nan2,
+      abs(vec1 - vec2) <= abs_tol,
+      if (na_equal) {
```

Comment:
doesn't this break the entire point of `fifelse` by forcing it to use an R for loop?